### PR TITLE
feat: add `process.getWindowsVersionType()` API

### PR DESCRIPTION
**Pull request by erickzhao**
#### Description of Change

Exposes a new API that returns Chromium's Windows `VersionType` enum.

https://source.chromium.org/chromium/chromium/src/+/master:base/win/windows_version.h;l=57-68;drc=76bb9de1de4e0e52d2d40c2c8d5641f7b20e4b50

cc @MarshallOfSound 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Added new `process.getWindowsVersionType()` API.

### CODE REVIEW

File: docs/api/process.md
Comment (MarshallOfSound) [inline docs/api/process.md:221]:
```suggestion
Returns `String` - Can be `server` or `desktop`. Returns `server` if operating system is Windows Server. Otherwise, returns `desktop`.
```

This should make the type definitions correct

Comment (erickzhao) [reply docs/api/process.md]: Thank you! Wasn't sure about the semantics. 😅

**Review by nornagon — CHANGES_REQUESTED** [review]
We should probably consider exposing the whole enum: https://source.chromium.org/chromium/chromium/src/+/master:base/win/windows_version.h;l=57-68;drc=76bb9de1de4e0e52d2d40c2c8d5641f7b20e4b50

I think this could also be named more specifically as it's only useful on Windows. Perhaps `getWindowsVersionType()`? I think it might also make sense to throw on non-Windows platforms.

**Comment by nornagon** [comment]
Also, what's the intended use case for this? What do apps need to differently on Windows Server than other versions of Windows?

**Comment by MarshallOfSound** [comment]
> Also, what's the intended use case for this? What do apps need to differently on Windows Server than other versions of Windows?

Sometimes when stopping support of older OS's you need to be able to distinguish between the same OS version running on different "product types".  i.e. Windows 7 Home vs Server 2008.  It's helpful information to send with autoUpdate requests.

**Comment by erickzhao** [comment]
@MarshallOfSound Is there a usecase to eventually expand this `server` vs `desktop` difference to Linux? In that case, it'd make sense for the API to not be renamed to be Windows-specific.

**Comment by nornagon** [comment]
> @MarshallOfSound Is there a usecase to eventually expand this `server` vs `desktop` difference to Linux? In that case, it'd make sense for the API to not be renamed to be Windows-specific.

I don't think it's a good idea to try to pretend that these things are comparable between linux and windows.

**Comment by MarshallOfSound** [comment]
Agreed, if we need the distinction on Linux it will probably have a different usecase and a vastly more subtle implementation.  Let's make this Windows specific for now

File: docs/api/process.md
Comment (nornagon) [inline docs/api/process.md:219]:
```suggestion
### `process.getWindowsVersionType()` _Windows_
```

File: shell/common/api/electron_bindings.cc
Comment (nornagon) [inline shell/common/api/electron_bindings.cc:221]: I think this would be better expressed with a `switch` statement.

File: spec/api-process-spec.js
Comment (nornagon) [inline spec/api-process-spec.js:77]:
```suggestion
  ifdescribe(process.platform === 'win32')('process.getWindowsVersionType()', () => {
```

& then drop the `if`s on the individual tests

**Review by nornagon — COMMENTED** [review]
API LGTM, couple of nits in the code.

**Review by nornagon — APPROVED** [review]
marking approved to speed things along, but please 👀  my comments before merging!

**Comment by jkleinsc** [comment]
API LGTM

File: spec/api-process-spec.js
Comment (erickzhao) [reply spec/api-process-spec.js:77]: The second test checks if it throws on non-Windows platforms, though. I copied the `ifit` pattern from other tests that tested platform-specific functionality.🤔

**Review by jkleinsc — CHANGES_REQUESTED** [review]
Blocking this until https://github.com/electron/electron/pull/28481/files#r696118830 is resolved.

**Comment by miniak** [comment]
@erickzhao are you still planning to land this PR?

File: shell/common/api/electron_bindings.cc
Comment (miniak) [inline shell/common/api/electron_bindings.cc:8]: I don't see an `std::map` used anywhere in your code

**Comment by miniak** [comment]
Agreed with @erickzhao to close the PR for now

**Pull request closed without merging**
